### ISSUE TITLE
bug(service-worker): does not start/install (zone not stabilizing)

### ISSUE THREAD

**Issue by Ploppy3** [angular/angular#20970]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[X] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
Service worker does not start/install.

## Expected behavior
Service worker should always start/install.

## Minimal reproduction of the problem with instructions

- create new project 
`ng new test-service-worker --service-worker`
- add this code to the app.component's constructor:
`setInterval(function() { console.log('test'); }, 1000);`
- build app 
`ng build --prod`
- service worker will never start/install

## Environment
<pre><code>
Angular version: 5.1.0
CLI: 1.6.0
</code></pre>

**Comment by Ploppy3** [comment]
@alxhub Hello, do you have any ETA on this bug? 

I'm also curious to know why using a `setInterval` could cause this issue. You may also look at the `setTimeout` function.

I have another project which has the a similar problem but does not use any `setInterval` as far as I know (maybe in the dependencies though). So it may be a wider problem.

**Comment by alxhub** [comment]
Hello @Ploppy3,

There is an ongoing investigation into this issue (which affects far more than service worker - the Angular zone never becomes stable in such an app).

**Comment by skydever** [comment]
I run into the same problem when I use `interval` from 
`import { interval } from 'rxjs/observable/interval';`
like in the service worker [docs](https://angular.io/guide/service-worker-communications#checking-for-updates).

As a workaround I did that

```javascript
this.ngZone.runOutsideAngular(() => {
  interval(ms).subscribe(() => {
    this.ngZone.run(() => this.updates.checkForUpdate());
  });
});
```

and everything works fine now, but this is just a workaround.

By the way, awesome service worker integration, **big thx** for that!

**Comment by Ploppy3** [comment]
@sky

**Comment by Ploppy3** [comment]
Th qsdqaaqsb

**Comment by Ploppy3** [comment]
@skydever I implemented this too but with the regular `setInterval` function.

**Comment by skydever** [comment]
@Ploppy3 ok, good to know

seems to be related to [#8089](https://github.com/angular/angular/issues/8089), so I am not sure if it really is just a workaround @alxhub ?

**Comment by Splaktar** [comment]
I have been fighting this issue for ~3 months as well. My NGSW never registers and it has been almost impossible to debug.

The information here is very helpful though. I run [FullStory](https://www.fullstory.com/) on my site and it is extremely important to keep in place. However, it does send a `POST` request every few seconds. You can observe this on https://www.marketamplified.com/ using Chrome's Network tab. This would very much explain the issue that I am seeing where the NGSW never gets registered, no matter what.

I'm working on verifying this atm. I also use Firebase Auth and Firebase DB, so those may also do something to keep the page from being 'stable'.

**Comment by Splaktar** [comment]
OK, I just verified that I had to disable both Firebase and Fullstory to get NGSW to register. After all of these months, just seeing it register finally was a great feeling!

Of course, I can't just pull Firebase and Fullstory from my app in order to use NGSW. Is there a param that I can sent to override this behavior and force registration?

**Comment by Splaktar** [comment]
As a workaround, I was able to modify some of the files in `node_modules/@angular/service-worker` to just do:
`return navigator.serviceWorker.register(script, { scope: options.scope });`

Then I was able to get my app working in addition to NGSW registered. 

It would be nice to add a `RegistrationOptions` field that allowed indicating the max time for the NGSW to wait for registration.

**Comment by Splaktar** [comment]
@alxhub I've created a reproduction repo [here](https://github.com/DevIntent/angular-example). You can see that the NGSW doesn't register [here](https://devintent-angular-example.firebaseapp.com).

**Comment by JiaLiPassion** [comment]
@Splaktar ,  I think maybe `zone.js/dist/zone-patch-rxjs` may resolve such kind of issue, I will test the repo you provided.

**Comment by JiaLiPassion** [comment]
@Ploppy3 , could you create a reproduce repo? I would like to debug the `zone` related part.

**Comment by Ploppy3** [comment]
@JiaLiPassion is it what you ask for? https://github.com/Ploppy3/zone-debug-service-worker

**Comment by JiaLiPassion** [comment]
@Ploppy3 , thank you, I think I may have found the reason, I will try to check it further and create a PR in `zone.js`.

**Comment by JiaLiPassion** [comment]
@Ploppy3 , I have created a PR, https://github.com/angular/zone.js/pull/999, I used you repo to test, it worked.

@Splaktar , could you try the attached `zone.js.zip` to test your repo? I used your repo and some `google api` error occurs.
[zone.js.zip](https://github.com/angular/angular/files/1664151/zone.js.zip)

**Comment by Ploppy3** [comment]
@JiaLiPassion Did you try with setTimeout? It may be a problem too, I did not test though.

**Comment by JiaLiPassion** [comment]
@Ploppy3 , I will test it, but `setTimeout` is different with `setInterval`, I think it should be ok.

**Comment by JiaLiPassion** [comment]
@Ploppy3, I believe performance will be the same with current version. 
except in some rare case, which is there are lot of `setInterval` and `cancelInterval` calls in the logic. in such kind of case, there will be performance improvement because in the new PR, `setInterval` will not trigger `onHasTask` any more.

**Comment by Ploppy3** [comment]
@JiaLiPassion Ok, thank you for the explanations, much appreciated.

**Comment by Splaktar** [comment]
@JiaLiPassion I can't really help with the Google API error since I don't have any details to work with. Perhaps you weren't hosting the project at `localhost:4200/`? I've added `127.0.0.1:4200/` to the allowed list as well.

I tested by copying your zip file into my `node_modules/zone.js/dist/` folder, doing a prod build, and then serving via `http-server dist/ -p 4200`. I tried with a couple different versions of Chrome and the Service Worker did not register in any of my tests.

I've updated the repo to fix some minor issues and include steps for testing the Service Worker in the README.

If I go into `node_modules/@angular/service-worker/esm5/service-worker.js` and change:
```
whenStable.then(function () { return navigator.serviceWorker.register(script, { scope: options.scope }); });
```
To
```
navigator.serviceWorker.register(script, { scope: options.scope });
```
Then the Service Worker registers properly with or without your changes to zone.js.

**Comment by Splaktar** [comment]
@jamesdaniels was taking a look at this in November on the Firebase side in https://github.com/angular/angularfire2/issues/1347, but it doesn't look like any real progress towards a solution was made there.

**Comment by JiaLiPassion** [comment]
@Splaktar, got it, I will try your update repo to find out why zone is not stable.

**Comment by JiaLiPassion** [comment]
@Splaktar, I think the reason is `firebase auth` make a very very long `setTimeout` to do `proactive token refresh`.  (more than 200000ms), so maybe `service worker` need to provide a manually register way, I will do a little more research.

**Comment by btroncone** [comment]
I am hitting this issue in one of my apps as well. Service worker silently fails to register despite having the same configuration as other apps that work. I am not using AngularFire so I'm unsure the cause, any pointers while we wait for a complete solution to be merged?

**Comment by sean-nicholas** [comment]
Maybe you have somewhere an interval or a long running timeout. Nevertheless, my fix was to register the service worker by myself: `navigator.serviceWorker.register('/ngsw-worker.js')` Just check if you are in production, the browser supports sw and no sw is currently active.

**Comment by JiaLiPassion** [comment]
@btroncone, I have created a PR https://github.com/angular/angular/pull/21842 to support more `strategy`. but as @sean-nicholas said, you can register by yourself.

**Comment by btroncone** [comment]
@sean-nicholas @jiaLiPassion Thanks! At what point do you register? Could you share the snippet you have for manually registering (w/ checks)? 

Appreciate the help!

**Comment by sean-nicholas** [comment]
@btroncone sure. I have a ServiceWorkerService that is injected into my main component. Here is the code from the service's constructor:
```ts
if (environment.production && 'serviceWorker' in navigator) {
  navigator.serviceWorker.getRegistration()
    .then(active => !active && navigator.serviceWorker.register('/ngsw-worker.js'))
    .catch(console.error)
}
```

**Comment by btroncone** [comment]
@sean-nicholas Thanks a ton!

**Comment by btroncone** [comment]
@sean-nicholas If you don't mind me asking, how are you managing registering for available updates without the use of the `SwUpdate` service?

**Comment by sean-nicholas** [comment]
@btroncone I'm using the `SwUpdate` as describes in the docs. You can still use `SwUpdate` after registering the sw yourself. It just works as expected.

**Comment by btroncone** [comment]
@sean-nicholas So you still include `ServiceWorkerModule.register(script, options)` in your main app module for the rest of the services? Even though the registration of the script fails the rest works perfectly? 

Thanks for your help!

**Comment by sean-nicholas** [comment]
@btroncone Correct :)

**Comment by Tecaa** [comment]
I have the same problem =(, I spent all my day searching what is the problem, at the end of the day I discover that it was the setInterval function in one of my components. this does not allow to serviceWorker to start.

The error that Google Chrome showed me in console was: "Site cannot be installed: no matching service worker detected. You may need to reload the page, or check that the service worker for the current page also controls the start URL from the manifest."

**Comment by rahulsahay19** [comment]
Hi, 

In my angular 6 app, after applying ng add @angular/pwa, it added all the required components. However, when I am running my app, service worker isn't getting registered. Here server is nodejs server. I am not serving via ng serve. problem which is happening is ngsw-worker.js file is not getting fetched. Although all related files are there in dist folder. I have attached screenshots for the same.

![1st](https://user-images.githubusercontent.com/3886381/42074970-5ad010dc-7b8c-11e8-965c-3cbb49a3bf80.png)

![2nd](https://user-images.githubusercontent.com/3886381/42074971-5afcb51a-7b8c-11e8-9ef8-57cce3ad4698.png)

![3rd](https://user-images.githubusercontent.com/3886381/42074972-5b28f8b4-7b8c-11e8-93f5-f9abc5be7bb2.png)

![4th](https://user-images.githubusercontent.com/3886381/42074973-5b53c918-7b8c-11e8-8826-14ed4bc81d3d.png)

I also tried to register the same from the main.ts file like shown below, but that also doesn't help.

`platformBrowserDynamic().bootstrapModule(AppModule)
.then(()=>{
  if ('serviceWorker' in navigator) {
    navigator.serviceWorker.getRegistration()
     .then(active => !active && navigator.serviceWorker.register('/ngsw-worker.js'))
      .catch(console.error)
  }
})
.catch(err => console.log(err));`

And  currently my service worker looks like this in browser.

![image](https://user-images.githubusercontent.com/3886381/42075080-d303f564-7b8c-11e8-8015-a391566777f5.png)

Kindly let me know if anything I am missing here? 

Thanks,
Rahul

**Comment by mackelito** [comment]
Did you follow the instructions on https://angular.io/guide/service-worker-getting-started ?

**Comment by rahulsahay19** [comment]
@mackelito Yes. I don't see any step missing here. Although, instead of http-server, I used my node js server for the same as app is getting served via that. However, my only concern is since our application is using signalr and it seems that there is some problem with signalr and websockets currently! Thanks!

**Comment by Spaier** [comment]
Hi,

In my angular 6 app service worker isn't registered when I enable services which are using `interval` function of `rxjs >= 6.0.0`.

Node: 8.11.3
Angular: 6.0.7
Angular CLI: 6.0.8
Server: ASP.NET Core 2.1

Service example:
```typescript
import { Injectable, Optional, OnDestroy } from '@angular/core'
import { SwUpdate } from '@angular/service-worker'

import { interval, Subject } from 'rxjs'
import { takeUntil } from 'rxjs/operators'

@Injectable({
  providedIn: 'root'
})
export class CheckForUpdateService implements OnDestroy {

  private readonly onDestroy = new Subject<void>()

  constructor(@Optional() private swUpdate: SwUpdate) {
    if (this.swUpdate && this.swUpdate.isEnabled) {
      const updateInterval = 1000 * 60 * 60 * 6 // 6 Hours
      // If I comment this code, service worker works.
      interval(updateInterval)
        .pipe(takeUntil(this.onDestroy))
        .subscribe(() => this.swUpdate.checkForUpdate())
    }
  }

  ngOnDestroy(): void {
    this.onDestroy.next()
    this.onDestroy.complete()
  }
}

```

**Comment by Delagen** [comment]
As I know services missing OnDestroy life hooks

**Comment by Spaier** [comment]
"Lifecycle hook that is called when a directive, pipe or service is destroyed" from [docs])(https://angular.io/api/core/OnDestroy)

**Comment by rahulsahay19** [comment]
@Spaier Hi, looks like your service worker with angular 6 is working fine. Can you have a look at my thread and point out if there is any bug or anything missing here.

My Build script goes like this.

  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build --prod --base-href /poseidon-home/",
    "ngsw-config": "node_modules/.bin/ngsw-config dist ngsw-config.json",
    "build-pwa": "npm run ngsw-config dist ngsw-config.json /base/href",
    "build-prod-ngsw": "ng build --prod --base-href /poseidon-home/ && npm run ngsw-config",
    "update": "copy /Y package.publish.json dist\\package.json",
    "test": "ng test --watch=false",
    "lint": "ng lint",
    "e2e": "ng e2e"
  }

app.module.ts
ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production })

However, I have optionally kept changes in main.ts file like

platformBrowserDynamic().bootstrapModule(AppModule).then(() => {
  if ('serviceWorker' in navigator && environment.production) {
    navigator.serviceWorker.register('/ngsw-worker.js');
  }
}).catch(err => console.log(err));

Even with this exclusive registration. It pops out the same issue.

Thanks!

**Comment by Spaier** [comment]
@rahulsahay19 Have you tried to generate and register service worker with Angular CLI? If you have any code with `interval` or `setInterval` it should be wrapped with `NgZone.RunOutsideAngular`.
My service worker configuration:
angular.browser.module.ts
```typescript
ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production })
```
main.ts
```typescript
document.addEventListener('DOMContentLoaded', () => {
  platformBrowserDynamic().bootstrapModule(AppBrowserModule).catch(err => console.log(err))
})
```
angular.json
```json
"production": {
    …
    "serviceWorker": true,
    // We can also specify ngsw path in angular.json
    // "ngswConfigPath": "src/ngsw-config.json"
}
```
package.json
```json
"build": "ng build --prod --ngsw-config-path src/ngsw-config.json && gulp",
```
gulpfile.js
```js
const gulp = require('gulp')
const minifyhtml = require('gulp-htmlmin')
const minifyjson = require('gulp-jsonminify')
const muxml = require('gulp-muxml')
const uglifyes = require('gulp-uglify-es').default
const pump = require('pump')

gulp.task('default', fixProdBuild)

function fixProdBuild(done) {
  let root = 'dist/vethelp'
  let output = 'dist/vethelp'
  pump([
    gulp.src([`${root}/ngsw-worker.js`, `${root}/safety-worker.js`, `${root}/worker-basic.min.js`]),
    uglifyes(),
    gulp.dest(output)
  ])
  pump([
    gulp.src(`${root}/**/*.json`),
    minifyjson(),
    gulp.dest(output)
  ])
  pump([
    gulp.src(`${root}/**/*.xml`),
    muxml({pretty: false}),
    gulp.dest(output)
  ])
  pump([
    gulp.src(`${root}/index.html`),
    minifyhtml({
      collapseWhitespace: true,
      removeComments: true
    }),
    gulp.dest(output)
  ], done())
}
```

**Comment by rahulsahay19** [comment]
@Spaier basically I used this "ng add @angular/pwa " drops all the required components and make all required changes in angular 6 app. 

Then, my package.json looks like

"ngsw-config": "node_modules/.bin/ngsw-config dist ngsw-config.json",
"build-pwa": "npm run ngsw-config dist ngsw-config.json /base/href",
"build-prod-ngsw": "ng build --prod --base-href /poseidon-home/ && npm run ngsw-config",

So, I am running the last one and deploying on server. Also, I have taken your main.ts code as well.

document.addEventListener('DOMContentLoaded', () => {
  platformBrowserDynamic().bootstrapModule(AppBrowserModule).catch(err => console.log(err))
})

But this also doesn't work and it expects one more thing start.html, this I saw in network tab. Hence flipped to 

platformBrowserDynamic().bootstrapModule(AppModule).then(() => {
  if ('serviceWorker' in navigator && environment.production) {
    navigator.serviceWorker.register('/ngsw-worker.js');
  }
}).catch(err => console.log(err));

Upon building, it drops all required components in my dist folder like shown below.

Here all the required components are there like service worker, safety worker, worker basic, manifest and ngsw.json.

![image](https://user-images.githubusercontent.com/3886381/42218165-4c06a9b0-7ee5-11e8-9813-53260e37d3a7.png)

However, my processed ngsw.json looks like

```json
{
  "configVersion": 1,
  "index": "/index.html",
  "assetGroups": [
    {
      "name": "app",
      "installMode": "prefetch",
      "updateMode": "prefetch",
      "urls": [
        "/favicon.ico",
        "/index.html",
        "/main.52e4d1b38b5feb7ea5ad.js",
        "/ngsw-worker.js",
        "/oidc-client.min.js",
        "/polyfills.855d06b353079dbdbf17.js",
        "/runtime.a66f828dca56eeb90e02.js",
        "/safety-worker.js",
        "/styles.9127785bd30cb224e73f.css",
        "/worker-basic.min.js"
      ],
      "patterns": []
    },
    {
      "name": "assets",
      "installMode": "lazy",
      "updateMode": "prefetch",
      "urls": [
        "/assets/icons/icon-128x128.png",
        "/assets/icons/icon-144x144.png",
        "/assets/icons/icon-152x152.png",
        "/assets/icons/icon-192x192.png",
        "/assets/icons/icon-384x384.png",
        "/assets/icons/icon-512x512.png",
        "/assets/icons/icon-72x72.png",
        "/assets/icons/icon-96x96.png",
        "/assets/img/components/logomark-kognifai-neg.svg",
        "/assets/img/components/logomark-sitecom-dv.png",
        "/assets/img/components/logotype-kognifai-neg.svg",
        "/assets/img/components/logotype-sitecom-dv.png",
        "/assets/img/gfx/example-icon.png",
        "/assets/img/gfx/kog-logo-neg.svg",
        "/assets/img/gfx/logotype-kognifai-neg.svg",
        "/assets/img/gfx/logotype-kognifai.svg",
        "/assets/img/icons/angle-down.svg",
        "/assets/img/icons/angle-left.svg",
        "/assets/img/icons/angle-right.svg",
        "/assets/img/icons/angle-swap-horizontal.svg",
        "/assets/img/icons/angle-swap-vertical.svg",
        "/assets/img/icons/angle-up.svg",
        "/assets/img/icons/arrow-down.svg",
        "/assets/img/icons/arrow-left.svg",
        "/assets/img/icons/arrow-right.svg",
        "/assets/img/icons/arrow-up.svg",
        "/assets/img/icons/attachment.svg",
        "/assets/img/icons/bookmark.svg",
        "/assets/img/icons/box.svg",
        "/assets/img/icons/caret-down.svg",
        "/assets/img/icons/caret-left.svg",
        "/assets/img/icons/caret-right.svg",
        "/assets/img/icons/caret-up.svg",
        "/assets/img/icons/check.svg",
        "/assets/img/icons/close.svg",
        "/assets/img/icons/cog.svg",
        "/assets/img/icons/danger.svg",
        "/assets/img/icons/dots-horizontal.svg",
        "/assets/img/icons/dots-vertical.svg",
        "/assets/img/icons/facebook.svg",
        "/assets/img/icons/industries/drilling-well-white.svg",
        "/assets/img/icons/industries/drilling-well.svg",
        "/assets/img/icons/industries/fishery-white.svg",
        "/assets/img/icons/industries/fishery.svg",
        "/assets/img/icons/industries/maritime-offshore-white.svg",
        "/assets/img/icons/industries/maritime-offshore.svg",
        "/assets/img/icons/industries/oil-gas-production-white.svg",
        "/assets/img/icons/industries/oil-gas-production.svg",
        "/assets/img/icons/industries/other-industries-white.svg",
        "/assets/img/icons/industries/other-industries.svg",
        "/assets/img/icons/industries/renewables-white.svg",
        "/assets/img/icons/industries/renewables.svg",
        "/assets/img/icons/industries/subsea-white.svg",
        "/assets/img/icons/industries/subsea.svg",
        "/assets/img/icons/industries/utilities-white.svg",
        "/assets/img/icons/industries/utilities.svg",
        "/assets/img/icons/info.svg",
        "/assets/img/icons/linkedin.svg",
        "/assets/img/icons/minus.svg",
        "/assets/img/icons/papers.svg",
        "/assets/img/icons/pin.svg",
        "/assets/img/icons/plus.svg",
        "/assets/img/icons/pull-down.svg",
        "/assets/img/icons/pull-left.svg",
        "/assets/img/icons/pull-right.svg",
        "/assets/img/icons/pull-up.svg",
        "/assets/img/icons/reload.svg",
        "/assets/img/icons/search.svg",
        "/assets/img/icons/sprites/icons.svg",
        "/assets/img/icons/success.svg",
        "/assets/img/icons/twitter.svg",
        "/assets/img/icons/user.svg",
        "/assets/img/icons/warning.svg"
      ],
      "patterns": []
    }
  ],
  "dataGroups": [],
  "hashTable": {
    "/assets/icons/icon-128x128.png": "dae3b6ed49bdaf4327b92531d4b5b4a5d30c7532",
    "/assets/icons/icon-144x144.png": "b0bd89982e08f9bd2b642928f5391915b74799a7",
    "/assets/icons/icon-152x152.png": "7479a9477815dfd9668d60f8b3b2fba709b91310",
    "/assets/icons/icon-192x192.png": "1abd80d431a237a853ce38147d8c63752f10933b",
    "/assets/icons/icon-384x384.png": "329749cd6393768d3131ed6304c136b1ca05f2fd",
    "/assets/icons/icon-512x512.png": "559d9c4318b45a1f2b10596bbb4c960fe521dbcc",
    "/assets/icons/icon-72x72.png": "c457e56089a36952cd67156f9996bc4ce54a5ed9",
    "/assets/icons/icon-96x96.png": "3914125a4b445bf111c5627875fc190f560daa41",
    "/assets/img/components/logomark-kognifai-neg.svg": "e84e077ec5ed3e8fbc9e172a4e6a5ff96cb93660",
    "/assets/img/components/logomark-sitecom-dv.png": "2a23fb8a3ab1ced65eb462c6f86e5304c26da54e",
    "/assets/img/components/logotype-kognifai-neg.svg": "99c0daf1d6f4f2f9ba794bf5c813c7fce281cc74",
    "/assets/img/components/logotype-sitecom-dv.png": "c8bc47d808f4513be8ddf1b73c3dbec8f9dd1286",
    "/assets/img/gfx/example-icon.png": "f363e3f58b0d8a547a7dadce5324d7b158823a39",
    "/assets/img/gfx/kog-logo-neg.svg": "f97b7851d5971530cb5a10d9b84be18aea7d93d9",
    "/assets/img/gfx/logotype-kognifai-neg.svg": "99c0daf1d6f4f2f9ba794bf5c813c7fce281cc74",
    "/assets/img/gfx/logotype-kognifai.svg": "59123fcdd06e588996b42add301f2dfc01198182",
    "/assets/img/icons/angle-down.svg": "adc49b0ac1731acc3b9cee2168757eca8b146c04",
    "/assets/img/icons/angle-left.svg": "879f0e9fb3d339873171952e7fead69990d85e04",
    "/assets/img/icons/angle-right.svg": "381025b855c9402e0bd95fdaf6b2da3feb2af6f0",
    "/assets/img/icons/angle-swap-horizontal.svg": "94162dd5de7c0b08e8a19e7b42faa7d0e0f4892f",
    "/assets/img/icons/angle-swap-vertical.svg": "b5f25d1f85bfd2e3a05785769d25c4ca484a4692",
    "/assets/img/icons/angle-up.svg": "6c7fae0d1e1d2924f6a1ccedea31804f0ed51907",
    "/assets/img/icons/arrow-down.svg": "bfb0df69db6074941c72651b0b90a0d6448d1035",
    "/assets/img/icons/arrow-left.svg": "f2c1fb923c6cb490aec3253715e2e501674a621d",
    "/assets/img/icons/arrow-right.svg": "df40c7bb1f690723abb80f60ccd08a99fac24a68",
    "/assets/img/icons/arrow-up.svg": "43c4bb42652f6b382c1ea261c5b92e8aedfc47b4",
    "/assets/img/icons/attachment.svg": "f1c1c29209e8421bc4a4be31a9a1fd5ac115ae34",
    "/assets/img/icons/bookmark.svg": "ca90113ff93dd39a594b208e9037e03614869afb",
    "/assets/img/icons/box.svg": "51139371653764351734343015b59636ab6e81fb",
    "/assets/img/icons/caret-down.svg": "2724eec85a42ec886a0d600c4d6b1a873422a50c",
    "/assets/img/icons/caret-left.svg": "e41f0f638958afcc88fed7570964aaf7ef1416ea",
    "/assets/img/icons/caret-right.svg": "73b72f26f190d39bc060eaf8eddc1afb5f62016c",
    "/assets/img/icons/caret-up.svg": "25bb6b07551a8408a3ad15f701fbacda74208063",
    "/assets/img/icons/check.svg": "05a0971fe92d98465a78a9c54157665a1058842a",
    "/assets/img/icons/close.svg": "702f1cea5ffb013897edf8851df9edb71867a009",
    "/assets/img/icons/cog.svg": "529847e4b502188226eca217d7d2e15d67df3545",
    "/assets/img/icons/danger.svg": "997f0c638d2d728c39fc7ad12f5fa104c76f7221",
    "/assets/img/icons/dots-horizontal.svg": "7ad335e87c857035f65372173fdcc9c1ce8d1ef5",
    "/assets/img/icons/dots-vertical.svg": "fb032577a3511e44d2dd4a263a18d3f244b466c6",
    "/assets/img/icons/facebook.svg": "7dbae1f4e48565a0ae4742570b27a9ebaecacdba",
    "/assets/img/icons/industries/drilling-well-white.svg": "cf503c61dd76f72a107e8bdd48e2a2fae3f512e1",
    "/assets/img/icons/industries/drilling-well.svg": "9b86a849a0fffb5ca48a690170bb77b3bbd61f38",
    "/assets/img/icons/industries/fishery-white.svg": "cb54b0054237f5b3396d2ce66c64f050469679d5",
    "/assets/img/icons/industries/fishery.svg": "38b0aba94f5b7be6bca009a99694f9079f106298",
    "/assets/img/icons/industries/maritime-offshore-white.svg": "a6d4a13485ede18de12c22e1d7ecf281c729ae48",
    "/assets/img/icons/industries/maritime-offshore.svg": "aef7beb3dc095e5ed7ed716d5ba086131c820990",
    "/assets/img/icons/industries/oil-gas-production-white.svg": "7c578ad84abf7f28f55884f77536879b8afdf0e7",
    "/assets/img/icons/industries/oil-gas-production.svg": "74d59a59dde893f5d86bf6ca2a5bd85b8a3ba47f",
    "/assets/img/icons/industries/other-industries-white.svg": "38a8f65d5cff99d84c68b105ff7cd2f756cef877",
    "/assets/img/icons/industries/other-industries.svg": "a5be386e5e84c64a728249f5293dbd27d484165e",
    "/assets/img/icons/industries/renewables-white.svg": "bebbdabf32e49a449a003298e5044db0e624bef7",
    "/assets/img/icons/industries/renewables.svg": "1823ef80e737babd3ba9fe69b6cf40809b6cdb31",
    "/assets/img/icons/industries/subsea-white.svg": "d09d218f5569aa3b91a334696c2d075d39b21eb5",
    "/assets/img/icons/industries/subsea.svg": "0c92e131cbd845e93a3cfa5cbe2a9550111401d8",
    "/assets/img/icons/industries/utilities-white.svg": "2791972a701558a201df41c8341c55aa223efa35",
    "/assets/img/icons/industries/utilities.svg": "18eb1184965904287607a2e2583e307e99e7a7d1",
    "/assets/img/icons/info.svg": "b645e5ec49d1ab39d6206681335ddd9b730cb765",
    "/assets/img/icons/linkedin.svg": "78db9c5e5865428731887f2e1e2fd993075e32c9",
    "/assets/img/icons/minus.svg": "325997f29ec5819f52c3916c392ac6d0de4a5557",
    "/assets/img/icons/papers.svg": "4762b3ca14bbc76749b9a0860353d57ec7807341",
    "/assets/img/icons/pin.svg": "61c86d54dbdb5022a9098c5b00a083d27bd656d9",
    "/assets/img/icons/plus.svg": "820bc97e222bb48967e0f7930b551db83692e9a6",
    "/assets/img/icons/pull-down.svg": "0c4dfd1551814a4e65da91ea834417013066eb96",
    "/assets/img/icons/pull-left.svg": "7522ba7cdd6daceb572fe5b1e361b7baeb9fb1da",
    "/assets/img/icons/pull-right.svg": "5a1468a7ae101f814f09f31eb31c5623d17d05db",
    "/assets/img/icons/pull-up.svg": "ea01631c8906683dc78e6c976ebe2ef2049de6c3",
    "/assets/img/icons/reload.svg": "b0b612d9c9c0da5b41b40dcde02de098e4ad59fc",
    "/assets/img/icons/search.svg": "2f6c7c8f5a4291b1154029cd284ff4a5ccdf2504",
    "/assets/img/icons/sprites/icons.svg": "d91df6aedf3bc8cbe4c34da4176cf51c834f5abf",
    "/assets/img/icons/success.svg": "4036cfad80bc2265232b3fde681685aa9cd54f6e",
    "/assets/img/icons/twitter.svg": "c7e2193fe891d359bbde2738e9e10bfa9eb8c030",
    "/assets/img/icons/user.svg": "865a0c668058de9d3abf30490204093387bd70ab",
    "/assets/img/icons/warning.svg": "4ccc7a17d76d7a43b30392b280b6e188492672fc",
    "/favicon.ico": "4176dbd696e347e7e31164bb326710ae2a67b52a",
    "/index.html": "878a49cce6acef88f69520bda44283cf582ed297",
    "/main.52e4d1b38b5feb7ea5ad.js": "e843388c8ecdac8dab560054c31833a7eead7a2c",
    "/ngsw-worker.js": "c5a902eadafef027da61d8960343491e15191ab4",
    "/oidc-client.min.js": "b518cec5a387b74ace4fa9f32c34c74e097ed4dd",
    "/polyfills.855d06b353079dbdbf17.js": "80da53373afccfd9904381e92c5a2c9440468f05",
    "/runtime.a66f828dca56eeb90e02.js": "078e320cc6fdaf355836c3b1c52b059cdd33fc7e",
    "/safety-worker.js": "ae5c8be3955e38edeeec09ed94b63d1e57f003c9",
    "/styles.9127785bd30cb224e73f.css": "a67ca82266ba439197e0b4ee264b57353b6679b6",
    "/worker-basic.min.js": "ae5c8be3955e38edeeec09ed94b63d1e57f003c9"
  },
  "navigationUrls": [
    {
      "positive": true,
      "regex": "^\\/.*$"
    },
    {
      "positive": false,
      "regex": "^\\/(?:.+\\/)?[^\\/]*\\.[^\\/]*$"
    },
    {
      "positive": false,
      "regex": "^\\/(?:.+\\/)?[^\\/]*__[^\\/]*$"
    },
    {
      "positive": false,
      "regex": "^\\/(?:.+\\/)?[^\\/]*__[^\\/]*\\/.*$"
    }
  ]
}
```
And with these changes, when I see my developer tool, I see ngsw-worker.js is giving 404 issue.

![image](https://user-images.githubusercontent.com/3886381/42218440-3731b7d6-7ee6-11e8-981f-ada693a25e31.png)

![image](https://user-images.githubusercontent.com/3886381/42218471-4fb29a6e-7ee6-11e8-9723-d081730f033e.png)

![image](https://user-images.githubusercontent.com/3886381/42218582-c12e4b02-7ee6-11e8-8041-179a080f1b84.png)

I also see you are using gulp. In my case, its just the CLI managed this internally with webpack. Here, wanted to understand whether in your network tab whether you see ngsw-worker.js file.

Thanks for your time!

**Comment by Spaier** [comment]
@rahulsahay19 I'd try to remove all modules from app.module, core.module and shared.module to see if it works with a blank application then add them one by one. Also `ngsf-config` binary isn't used in angular cli build with service worker, have you tried to specify `--ngsw-config-path your/path/ngsw-config.json` and remove `ngsf-config` calls?
![2018-07-03 1](https://user-images.githubusercontent.com/18536847/42219491-beab69f8-7ed4-11e8-8a60-6eca52e6cead.png)
![2018-07-03](https://user-images.githubusercontent.com/18536847/42219492-becfbd08-7ed4-11e8-83d5-af67aad3ce85.png)

**Comment by rahulsahay19** [comment]
@Spaier Thanks for your response. It got fixed. Basically the problem was with base-href setting. so, my ngsw-worker.js sits after localhost/site/. And 

```ts
ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production })
```
 takes actual path than relative path. hence made the same relative by just putting 
```ts
ServiceWorkerModule.register('./ngsw-worker.js', { enabled: environment.production })
```
 and also modified manifest.json for relative path like shown below.

```json
{
  "name": "kognifai-poseidon-home",
  "short_name": "kognifai-poseidon-home",
  "theme_color": "#1976d2",
  "background_color": "#fafafa",
  "display": "standalone",
  "scope": "./",
  "start_url": "./",
  "icons": [
    {
      "src": "assets/icons/icon-72x72.png",
      "sizes": "72x72",
      "type": "image/png"
    },
    {
      "src": "assets/icons/icon-96x96.png",
      "sizes": "96x96",
      "type": "image/png"
    },
    {
      "src": "assets/icons/icon-128x128.png",
      "sizes": "128x128",
      "type": "image/png"
    },
    {
      "src": "assets/icons/icon-144x144.png",
      "sizes": "144x144",
      "type": "image/png"
    },
    {
      "src": "assets/icons/icon-152x152.png",
      "sizes": "152x152",
      "type": "image/png"
    },
    {
      "src": "assets/icons/icon-192x192.png",
      "sizes": "192x192",
      "type": "image/png"
    },
    {
      "src": "assets/icons/icon-384x384.png",
      "sizes": "384x384",
      "type": "image/png"
    },
    {
      "src": "assets/icons/icon-512x512.png",
      "sizes": "512x512",
      "type": "image/png"
    }
  ]
}

```
Thanks,
Rahul

**Comment by EatonZ** [comment]
Ran into this issue on a new Angular 6.1.4 app. Service Worker doesn't install when rxjs' `interval` is used as advised in the [official documentation](https://angular.io/guide/service-worker-communications#checking-for-updates). When it's commented out, everything works as expected.

Is there any sign of an official workaround/fix for this issue? It seems odd to me that the docs would recommend this approach if it doesn't work. It seems like there should be some kind of note in the docs that warns about this behavior.

**Comment by smnbbrv** [comment]
This is sooooo ugly. Just spent two hours to find out that the documentation is recommending the not-working "solution".

I'm fine with running things outside angular zone, I don't even care whether this is a workaround or not. But, **for the god's sake**, I ask for all the people who will come to this issue some day, **fix the documentation!** 

Better it recommends nothing than something that is known to be not working since already half a year.

**Comment by Ploppy3** [comment]
@smnbbrv You can open an issue regarding the documentation.

Also, only the setTimeout or setInterval statements should run outside the zone, the given function should run in the zone and therefore it does not affect your app.

```
this.ngZone.runOutsideAngular(() => {
  interval(1000).subscribe(() => {
    this.ngZone.run(() => {
      // put your usual code here
    });
  });
});
```

**Comment by Teamop** [comment]
With code comes from [angular-cli@8779](https://github.com/angular/angular-cli/issues/8779#issuecomment-350737663), also works. That's, wrap the `interval` part after the application is stable. thanks @dirand
```
export class AppComponent implements OnInit {
  title = 'app';

  constructor(private applicationRef: ApplicationRef) {
  }

  ngOnInit() {

    this.applicationRef.isStable.subscribe((s) => { // #1
      if (s) { // #2
        setInterval(t => { console.log('Ping')}, 500);
      } // #3
    }); // #4
    // If you uncomment 1-4 - service-worker will not run

    this.applicationRef.isStable.subscribe(t => {
      console.log('App stable: ' + t);
    });
  }
}
```

**Comment by stefan-karlsson** [comment]
Should it be possible to throw a error if the application does not reach a stable state within x seconds after calling the .register() function? 

I think that would help a lot of users having issue with this.

**Comment by gkalpak** [comment]
Once #21842 lands, we could consider changing the default strategy to registering once stable or when a certain amont of time passes (whichever comes first).
